### PR TITLE
Define default profile for image #4567

### LINKED ISSUE #4567: Define default profile for image
When I launch a new container I can specify which profile it should take instead of the default profile with the `-p` option:

```sh
lxc launch [<remote>:]<image> -p <some profile>
```
This profile can contain some default network setup (bridge, devices, ...) and the container won't work properly if you don't launch it with the correct profile.

It would be nice if you could specify a custom default profile to take when launching a container from a specific image instead of the real default profile. This could look something like this:
```sh
lxc image edit <image>
```
With the option:
```yaml
# ...
properties:
  default_profile: <profile>
# ...
``` 
Of course it would be nice to also include this option in the [metadata.yaml file for importing an image](https://github.com/lxc/lxd/blob/master/doc/image-handling.md#content)

This topic was already mentioned in a [linux container discussion thread](https://discuss.linuxcontainers.org/t/use-a-remote-profile/544) but I wanted to know if there is anything planed like this.

### PR DESCRIPTION
**Pull request by ralubis**
Closes #4567

Signed-off-by: Jack Stenglein <jackstenglein@gmail.com>
Signed-off-by: Rizwan Lubis <rizwan.lubis@gmail.com>

### CODE REVIEW

**Comment by stgraber** [comment]
Looks like this needs rebasing to apply onto master

**Comment by stgraber** [comment]
Some of your commits are lacking the sign off line.

**Comment by ralubis** [comment]
Right, we are actively developing on the branch and will fix the sign-offs when we squash the commits. This is intended to be a draft pull request so that we can make sure that we are on the right track, as you suggested in the issue. Could you take a look at the code itself to see if our approaches seem correct? Thanks!

File: lxc/image.go [hunk 1]
```diff
@@ -406,6 +406,9 @@ func (c *cmdImageEdit) Run(cmd *cobra.Command, args []string) error {
 		newdata := api.ImagePut{}
 		err = yaml.Unmarshal(content, &newdata)
 		if err == nil {
+			if newdata.Profiles == nil {
+				newdata.Profiles = []string{"default"}
+			}
```
Comment (stgraber) [inline lxc/image.go]: This should be handled on the server side, not on the client side.

File: lxc/init.go [hunk 2]
```diff
@@ -219,16 +219,6 @@ func (c *cmdInit) create(conf *config.Config, args []string) (lxd.InstanceServer
...
-
-	if !c.flagNoProfiles && len(profiles) == 0 {
-		if len(stdinData.Profiles) > 0 {
-			req.Profiles = stdinData.Profiles
-		} else {
-			req.Profiles = nil
-		}
-	} else {
-		req.Profiles = profiles
-	}
```
Comment (stgraber) [inline lxc/init.go]: This also should be handled by the server rather than the client.
Effectively replacing our current logic that substitutes a `nil` by the default profile.

File: lxd/container.go [hunk 3]
```diff
@@ -836,7 +836,12 @@ func instanceCreateInternal(s *state.State, args db.InstanceArgs) (instance.Inst
 	}
 
 	if args.Profiles == nil {
-		args.Profiles = []string{"default"}
+		profiles, err := s.Cluster.ImageProfilesGet(args.BaseImage, args.Project)
+		if err != nil {
+			return nil, err
+		}
+		args.Profiles = profiles
+		logger.Infof("Got profiles: %v", args.Profiles)
```
Comment (stgraber) [inline lxd/container.go]: Logic looks good so with that in place, you shouldn't need that `lxc/init.go` change.

File: lxd/db/images.go [hunk 4]
```diff
@@ -499,7 +504,7 @@ func (c *Cluster) imageFill(id int, image *api.Image, create, expire, used, uplo
 	}
 
 	image.Aliases = aliases
-
+
```
Comment (stgraber) [inline lxd/db/images.go]: What's that doing here?

File: test/suites/image_profiles.sh [hunk 5]
```diff
@@ -0,0 +1,41 @@
+test_image_nil_profile_list() {
+	ensure_import_testimage
+	lxc launch testimage c1
+	lxc list | grep -q "c1" || false
+    lxc info c1 | grep -q "Profiles: default" || false
```
Comment (randombenj) [inline test/suites/image_profiles.sh]: Use a tab here instead of 4 spaces?

**Comment by stgraber** [comment]
Can you rebase the branch once you're happy with your edits?

It makes reviewing it a bit easier.

**Comment by jackstenglein** [comment]
@stgraber We think we are done, but we are failing some tests when we run them locally. We're not sure if this is because of some bug in our code/the test or if our machine is not configured correctly.

**Comment by stgraber** [comment]
Okay, lets see what this does on Jenkins then.

**Comment by stgraber** [comment]
jenkins: test this please

File: lxc/image.go [hunk 6]
```diff
@@ -939,6 +942,15 @@ func (c *cmdImageInfo) Run(cmd *cobra.Command, args []string) error {
 		fmt.Printf("    Alias: %s\n", info.UpdateSource.Alias)
 	}
 
+	if len(info.Profiles) == 0 {
+		fmt.Printf(i18n.G("Profiles: ") + "[]\n")
+	} else {
+		fmt.Println(i18n.G("Profiles:"))
```
Comment (stgraber) [inline lxc/image.go]: Those lines are causing an addition to the translation template, this then requires an additional commit which you can generate by doing:
 - make i18n
 - git commit -s -m "i18n: Update translation template" po/

**Comment by stgraber** [comment]
The other test failure suggests that after running your new test, we're ending up with a `testimage` which doesn't have the `default` profile tied to it, causing the test after it to fail.

Deleting the image at the end of your test should prevent this from happening.

**Comment by jackstenglein** [comment]
Thanks! I think we have it working now. 

jenkins: run it back

**Comment by stgraber** [comment]
jenkins: test this please

File: lxc/image.go [hunk 7]
```diff
@@ -379,6 +379,9 @@ func (c *cmdImageEdit) Run(cmd *cobra.Command, args []string) error {
 		if err != nil {
 			return err
 		}
+		if newdata.Profiles == nil {
+			newdata.Profiles = []string{"default"}
+		}
```
Comment (stgraber) [inline lxc/image.go]: This part really shouldn't be needed.

File: lxd/db/images.go [hunk 8]
```diff
@@ -7,12 +7,11 @@ import (
 	"fmt"
 	"time"
 
-	"github.com/pkg/errors"
-
 	"github.com/lxc/lxd/lxd/db/query"
 	"github.com/lxc/lxd/lxd/instance/instancetype"
 	"github.com/lxc/lxd/shared/api"
 	"github.com/lxc/lxd/shared/osarch"
+	"github.com/pkg/errors"
```
Comment (stgraber) [inline lxd/db/images.go]: This change is un-needed and incorrect, the current order is the expected one for LXD Go files.

File: lxd/container.go [hunk 9]
```diff
@@ -836,7 +836,15 @@ func instanceCreateInternal(s *state.State, args db.InstanceArgs) (instance.Inst
...
-		args.Profiles = []string{"default"}
+		if args.BaseImage == "" {
+			args.Profiles = []string{"default"}
+		} else {
+			profiles, err := s.Cluster.ImageProfilesGet(args.BaseImage, args.Project)
+			if err != nil {
+				return nil, err
+			}
+			args.Profiles = profiles
+		}
```
Comment (stgraber) [inline lxd/container.go]: This isn't the right location for this change. `BaseImage` is also used with operations like migration and your change will cause failures in some cases there.

Instead you want the image profile logic to be done in `instanceCreateFromImage`.

File: shared/version/api.go [hunk 10]
```diff
@@ -179,7 +179,7 @@ var APIExtensions = []string{
 	"container_syscall_intercept_mount_fuse",
 	"container_disk_ceph",
 	"virtual-machines",
-	"image_profiles"
+	"image_profiles",
```
Comment (stgraber) [inline shared/version/api.go]: This change should be in the `api` commit.

**Comment by stgraber** [comment]
Things are looking pretty good. I've left a few in-line comments above.

On top of that, it'd be good if you could:
 - Move CLI changes to a `lxc/image: Add support for image profiles` commit
 - Make sure all changes to `lxd/db` are in the db commit rather than mixed with the other `lxd` commit

**Comment by jackstenglein** [comment]
Container.go change has been moved to `instanceCreateFromImage` and commit history has been cleaned up.

File: test/includes/setup.sh [hunk 11]
```diff
@@ -23,7 +23,8 @@ ensure_import_testimage() {
                 exit 1
             fi
 
-            deps/import-busybox --alias testimage
+            project=$(lxc project list | grep '(current)' | cut -d' ' -f2)
+            deps/import-busybox --alias testimage --project "$project"
```
Comment (stgraber) [inline test/includes/setup.sh]: Can you move this change to its own commit as it's effectively a generic fix to `ensure_import_testimage` to respect projects.

**Comment by stgraber** [comment]
jenkins: test this please

**Comment by stgraber** [comment]
Jenkins seems happy with it. I did another round of review and this looks good to me. The tests cover all the cases we care about I think and they're passing.

Thanks!
